packages/internal/src/js-utils.js

```diff
@@ -87,3 +129,45 @@ export const makeMeasureSeconds = currentTimeMillisec => {
   };
   return measureSeconds;
 };
+
+/**
+ * Generalize Node.js util.promisify to support callbacks with multiple
+ * post-error parameters, which fulfill the returned promise as an array rather
+ * than as a single value.
+ *
+ * @template {(...args: any[]) => unknown} F function whose last parameter is an
+ *   error-first callback
+ * @template {Parameters<LastParameter<F>> extends [Error | null, ...infer R]
+ *     ? R
+ *     : never} R
+ * @param {F} fn
+ */
+export const promisify = fn => {
+  /**
+   * @typedef {Parameters<Overloads<F>> extends [...infer P, infer CB]
+   *     ? (
+   *         ...args: P
+   *       ) => Promise<
+   *         Parameters<LastParameter<F>> extends [unknown, unknown] ? R[0] : R
+   *       >
+   *     : never} Promisified
+   */
+  const promisified = /** @type {Promisified} */ (
+    (...args) =>
+      // eslint-disable-next-line no-restricted-syntax
+      new Promise((resolve, reject) => {
+        /** @type {(err: Error, ...result: R) => void} */
+        const callback = (err, ...result) => {
+          if (err) {
+            reject(err);
+          } else if (result.length === 1) {
+            resolve(/** @type {any} */ (result[0]));
+          } else {
+            resolve(result);
```

Comment:
I dislike changing the type of resolved values dynamically like that but I'll allow it here for backwards compat (and because I think the types get it).

But I can't shake the feeling this is trying to be too smart.